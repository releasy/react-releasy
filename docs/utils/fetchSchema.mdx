---
name: fetchSchema
menu: Utils
route: /utils/fetchSchema
---

# fetchSchema

## Usage

```javascript
import { fetchSchema } from 'react-releasy';

fetchSchema({
  url: 'https://yourserveraddress.com/graphql',
  dir: './data',
}).then(() => {
  console.log('Schema downloaded');
  process.exit();
});

```
