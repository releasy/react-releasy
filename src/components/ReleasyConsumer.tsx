import * as React from 'react';

import GlobalContext, { GlobalContextValueType } from '../contexts/GlobalContext';

type ReleasyConsumerPropsType = {
  children: React.ReactNode,
};

const ReleasyConsumer = ({
  children,
}: ReleasyConsumerPropsType): React.ReactElement<GlobalContextValueType> => (
  <GlobalContext.Consumer>
    {(globalContextValue: GlobalContextValueType) =>
      React.Children.map(children, (child: React.ReactElement<any>) =>
        React.cloneElement(child, globalContextValue),
      )
    }
  </GlobalContext.Consumer>
);

export default ReleasyConsumer;
