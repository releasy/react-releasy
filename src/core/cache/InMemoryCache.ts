import { Variables, GraphQLResponse } from 'relay-runtime';
import RelayQueryResponseCache from 'relay-runtime/lib/RelayQueryResponseCache.js';

import { CacheInterface } from '../Config';

type InMemoryCacheProps = {
  ttl?: number,
  size?: number,
};

class InMemoryCache implements CacheInterface {
  store: RelayQueryResponseCache;

  constructor(props?: InMemoryCacheProps) {
    this.store = new RelayQueryResponseCache({
      size: props && props.size ? props.size : 250,
      ttl: props && props.ttl ? props.ttl : 60 * 1000 * 5, // default = 5 minutes
    });
  }

  get = (queryID: string, variables: Variables): GraphQLResponse => {
    return this.store.get(queryID, variables);
  }

  set = (queryID: string, variables: Variables, payload: GraphQLResponse): void => {
    this.store.set(queryID, variables, payload);
  }

  clear = (): void => {
    this.store.clear();
  }
}

export default InMemoryCache;
