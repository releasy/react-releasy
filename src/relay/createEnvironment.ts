import { installRelayDevTools } from 'relay-devtools';
import { Environment, RecordSource, Store } from 'relay-runtime';

import Config from '../core/Config';
import createNetwork from './createNetwork';

const createEnvironment = (config: Config): Environment => {
  if (config.devTools) {
    installRelayDevTools();
  }

  const network = createNetwork(config);

  const source = new RecordSource();
  const store = new Store(source);

  return new Environment({
    network,
    store,
  });
};

export default createEnvironment;
