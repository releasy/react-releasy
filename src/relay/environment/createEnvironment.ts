import { installRelayDevTools } from 'relay-devtools';
import { Environment, RecordSource, Store } from 'relay-runtime';

import Config from '../../core/Config';
import createNetwork from '../network/createNetwork';

const createEnvironment = (config: Config): Environment => {
  if (config.devTools) {
    installRelayDevTools();
  }

  const network = createNetwork(config);

  const recordSource = new RecordSource();
  const store = new Store(recordSource);

  return new Environment({
    network,
    store,
  });
};

export default createEnvironment;
