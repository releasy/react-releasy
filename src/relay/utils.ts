import { CacheConfig, RequestNode } from 'relay-runtime';

export const isMutation = (request: RequestNode) => request.operationKind === 'mutation';

export const mustForceFetch = (cacheConfig: CacheConfig) => !!(cacheConfig && cacheConfig.force);
