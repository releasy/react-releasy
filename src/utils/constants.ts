export const IS_NODE = typeof window === 'undefined';

export const WINDOW_DATA_KEY = '__RELEASY_DATA__';
