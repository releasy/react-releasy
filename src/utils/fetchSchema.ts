import {
  buildClientSchema,
  printSchema,
  introspectionQuery,
} from 'graphql';
import fse from 'fs-extra';
import fetch from 'node-fetch';

type FetchSchemaInput = {
  url: string,
  path: string,
};

const fetchSchema = async (input: FetchSchemaInput): Promise<void> => {
  const { url, path } = input;

  const response = await fetch(url, {
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      query: introspectionQuery,
    }),
  });

  const result = await response.json();

  const clientSchema = buildClientSchema(result.data);
  const schema = printSchema(clientSchema);

  fse.ensureFile(path);

  return fse.writeFile(path, schema);
};

export default fetchSchema;
