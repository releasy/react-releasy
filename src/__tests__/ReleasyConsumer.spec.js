import * as React from 'react';
import { mount, shallow } from 'enzyme';

import DumbComponent from '../../test/components/DumbComponent';
import createWithMutation from '../../test/components/createWithMutation';
import createWithRelayQueryRenderer from '../../test/components/createWithRelayQueryRenderer';
import mockFetch from '../../test/fixtures/fetch/mockFetch';
import createConfig from '../../test/fixtures/createConfig';
import { CONTENT_TYPE, MOCK_TYPE } from '../../test/consts';
import { delay } from '../../test/utils';

import InMemoryCache from '../cache/InMemoryCache';
import createEnvironment from '../relay/createEnvironment';
import ReleasyConsumer from '../ReleasyConsumer';

it('should render a valid ReleasyConsumer', () => {
  const config = createConfig();
  const environment = createEnvironment(config);

  const component = shallow(
    <ReleasyConsumer>
      <DumbComponent />
    </ReleasyConsumer>,
    { context: { environment } },
  );

  expect(component.instance()).toMatchSnapshot();
});

it('should thrown an error without "children"', () => {
  const renderConsumer = () => {
    shallow(<ReleasyConsumer />);
  };

  expect(renderConsumer).toThrowErrorMatchingSnapshot();
});

it('should render a loading QueryRenderer', async () => {
  const WithRelayQueryRenderer = createWithRelayQueryRenderer();

  const config = createConfig();
  const environment = createEnvironment(config);

  mockFetch();

  const component = mount(
    <ReleasyConsumer>
      <WithRelayQueryRenderer />
    </ReleasyConsumer>,
    { context: { environment } },
  );

  expect(component.find(WithRelayQueryRenderer).text()).toMatchSnapshot();
});

it('should render a success QueryRenderer', async () => {
  const WithRelayQueryRenderer = createWithRelayQueryRenderer();

  const config = createConfig();
  const environment = createEnvironment(config);

  mockFetch();

  const component = mount(
    <ReleasyConsumer>
      <WithRelayQueryRenderer />
    </ReleasyConsumer>,
    { context: { environment } },
  );

  await delay(100);

  expect(component.find(WithRelayQueryRenderer).text()).toMatchSnapshot();
});

it('should render an error QueryRenderer', async () => {
  const WithRelayQueryRenderer = createWithRelayQueryRenderer();

  const config = createConfig();
  const environment = createEnvironment(config);

  // let's set a different content type and payload body to see it breaking muhahaha (666)
  mockFetch({ contentType: CONTENT_TYPE.ALL, mock: MOCK_TYPE.ME_QUERY });

  const component = mount(
    <ReleasyConsumer>
      <WithRelayQueryRenderer />
    </ReleasyConsumer>,
    { context: { environment } },
  );

  await delay(100);

  expect(component.find(WithRelayQueryRenderer).text()).toMatchSnapshot();
});

it('should store data in the cache', async () => {
  const WithRelayQueryRenderer = createWithRelayQueryRenderer();

  const cache = new InMemoryCache();
  const config = createConfig({ cache });
  const environment = createEnvironment(config);

  mockFetch();

  mount(
    <ReleasyConsumer>
      <WithRelayQueryRenderer />
    </ReleasyConsumer>,
    { context: { environment } },
  );

  await delay(100);

  expect(config.cache.store._responses.size).toBe(1);
});

it('should hit cache', async () => {
  const WithRelayQueryRenderer = createWithRelayQueryRenderer();

  const cache = new InMemoryCache();
  const config = createConfig({ cache });
  const environment = createEnvironment(config);

  mockFetch();

  mount(
    <ReleasyConsumer>
      <WithRelayQueryRenderer />
    </ReleasyConsumer>,
    { context: { environment } },
  );

  await delay(100);

  mount(
    <ReleasyConsumer>
      <WithRelayQueryRenderer />
    </ReleasyConsumer>,
    { context: { environment } },
  );

  await delay(100);

  expect(config.cache.store._responses.size).toBe(1);
});

it('should clear cache', async () => {
  const name = 'Updated John Doe';

  const WithMutation = createWithMutation({ name });
  const WithRelayQueryRenderer = createWithRelayQueryRenderer();

  const cache = new InMemoryCache();
  const config = createConfig({ cache });
  const environment = createEnvironment(config);

  mockFetch();

  mount(
    <ReleasyConsumer>
      <WithRelayQueryRenderer />
    </ReleasyConsumer>,
    { context: { environment } },
  );

  await delay(100);
  mockFetch({
    mock: MOCK_TYPE.CHANGE_NAME_MUTATION,
    data: { name },
  });

  mount(
    <ReleasyConsumer>
      <WithMutation />
    </ReleasyConsumer>,
    { context: { environment } },
  );

  await delay(100);

  expect(config.cache.store._responses.size).toBe(0);
});
