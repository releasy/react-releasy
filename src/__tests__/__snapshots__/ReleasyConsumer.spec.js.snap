// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a loading QueryRenderer 1`] = `"loading"`;

exports[`should render a success QueryRenderer 1`] = `"fakeViewerId"`;

exports[`should render a valid ReleasyConsumer 1`] = `
ReleasyConsumer {
  "context": Object {
    "environment": RelayModernEnvironment {
      "__setNet": [Function],
      "_deferrableSelections": Set {},
      "_network": Object {
        "execute": [Function],
      },
      "_publishQueue": RelayPublishQueue {
        "_appliedOptimisticUpdates": Set {},
        "_backup": Object {},
        "_handlerProvider": [Function],
        "_pendingBackupRebase": false,
        "_pendingData": Set {},
        "_pendingOptimisticUpdates": Set {},
        "_pendingUpdaters": Set {},
        "_store": RelayMarkSweepStore {
          "_gcEnabled": true,
          "_hasScheduledGC": false,
          "_index": 0,
          "_recordSource": Object {},
          "_roots": Map {},
          "_subscriptions": Set {},
          "_updatedRecordIDs": Object {},
        },
      },
      "_store": RelayMarkSweepStore {
        "_gcEnabled": true,
        "_hasScheduledGC": false,
        "_index": 0,
        "_recordSource": Object {},
        "_roots": Map {},
        "_subscriptions": Set {},
        "_updatedRecordIDs": Object {},
      },
      "configName": undefined,
      "unstable_internal": Object {
        "areEqualSelectors": [Function],
        "createFragmentSpecResolver": [Function],
        "createOperationSelector": [Function],
        "getDataIDsFromObject": [Function],
        "getFragment": [Function],
        "getRequest": [Function],
        "getSelector": [Function],
        "getSelectorList": [Function],
        "getSelectorsFromObject": [Function],
        "getVariablesFromObject": [Function],
      },
    },
  },
  "props": Object {
    "children": <DumbComponent />,
  },
  "refs": Object {},
  "state": Object {},
  "updater": Updater {
    "_callbacks": Array [],
    "_renderer": ReactShallowRenderer {
      "_context": Object {
        "environment": RelayModernEnvironment {
          "__setNet": [Function],
          "_deferrableSelections": Set {},
          "_network": Object {
            "execute": [Function],
          },
          "_publishQueue": RelayPublishQueue {
            "_appliedOptimisticUpdates": Set {},
            "_backup": Object {},
            "_handlerProvider": [Function],
            "_pendingBackupRebase": false,
            "_pendingData": Set {},
            "_pendingOptimisticUpdates": Set {},
            "_pendingUpdaters": Set {},
            "_store": RelayMarkSweepStore {
              "_gcEnabled": true,
              "_hasScheduledGC": false,
              "_index": 0,
              "_recordSource": Object {},
              "_roots": Map {},
              "_subscriptions": Set {},
              "_updatedRecordIDs": Object {},
            },
          },
          "_store": RelayMarkSweepStore {
            "_gcEnabled": true,
            "_hasScheduledGC": false,
            "_index": 0,
            "_recordSource": Object {},
            "_roots": Map {},
            "_subscriptions": Set {},
            "_updatedRecordIDs": Object {},
          },
          "configName": undefined,
          "unstable_internal": Object {
            "areEqualSelectors": [Function],
            "createFragmentSpecResolver": [Function],
            "createOperationSelector": [Function],
            "getDataIDsFromObject": [Function],
            "getFragment": [Function],
            "getRequest": [Function],
            "getSelector": [Function],
            "getSelectorList": [Function],
            "getSelectorsFromObject": [Function],
            "getVariablesFromObject": [Function],
          },
        },
      },
      "_element": <ReleasyConsumer>
        <DumbComponent />
      </ReleasyConsumer>,
      "_forcedUpdate": false,
      "_instance": [Circular],
      "_newState": null,
      "_rendered": Array [
        <DumbComponent
          environment={
            RelayModernEnvironment {
              "__setNet": [Function],
              "_deferrableSelections": Set {},
              "_network": Object {
                "execute": [Function],
              },
              "_publishQueue": RelayPublishQueue {
                "_appliedOptimisticUpdates": Set {},
                "_backup": Object {},
                "_handlerProvider": [Function],
                "_pendingBackupRebase": false,
                "_pendingData": Set {},
                "_pendingOptimisticUpdates": Set {},
                "_pendingUpdaters": Set {},
                "_store": RelayMarkSweepStore {
                  "_gcEnabled": true,
                  "_hasScheduledGC": false,
                  "_index": 0,
                  "_recordSource": Object {},
                  "_roots": Map {},
                  "_subscriptions": Set {},
                  "_updatedRecordIDs": Object {},
                },
              },
              "_store": RelayMarkSweepStore {
                "_gcEnabled": true,
                "_hasScheduledGC": false,
                "_index": 0,
                "_recordSource": Object {},
                "_roots": Map {},
                "_subscriptions": Set {},
                "_updatedRecordIDs": Object {},
              },
              "configName": undefined,
              "unstable_internal": Object {
                "areEqualSelectors": [Function],
                "createFragmentSpecResolver": [Function],
                "createOperationSelector": [Function],
                "getDataIDsFromObject": [Function],
                "getFragment": [Function],
                "getRequest": [Function],
                "getSelector": [Function],
                "getSelectorList": [Function],
                "getSelectorsFromObject": [Function],
                "getVariablesFromObject": [Function],
              },
            }
          }
        />,
      ],
      "_rendering": false,
      "_updater": [Circular],
    },
  },
}
`;

exports[`should render an error QueryRenderer 1`] = `"self.body.on is not a function"`;

exports[`should thrown an error without "children" 1`] = `"ReleasyConsumer: missing \\"children\\""`;
